design-proposals/e2e-waitfor-tooling.md

```diff
@@ -0,0 +1,154 @@
+  the polling approach.
+- Increase e2e test stabilization.
+
+## Non Goals
+- Replace all existing polling implementation at once.
+
+# Proposal
+
+## Definition of Personas
+
+- E2E test authors and reviewers.
+
+## User Stories
+
+As a e2e test author:
+
+I want to execute an asynchronous operation on the cluster (e.g. create
+and run a VMI) and wait for it to complete, given a condition I can
+define and express (e.g. VMI phase `running`).
+I am able to add a timeout and abort the "waiting" explicitly.
+
+## Existing solution
+
+The current e2e tests at kubevirt/kubevirt use several methods to
+implement the "waiting" part, all are based on polling:
+- Ginkgo [Eventually](https://onsi.github.io/ginkgo/#patterns-for-asynchronous-testing).
+  Repeatedly executing the body function until the resulting assertion
+  succeeds, the timeout is reached or a body assertion fails.
+
+  Example:
+  ```
+  Eventually(func() error {
+      return libnet.PingFromVMConsole(vmi, ipAddress)
+  }, 15*time.Second, time.Second).Should(Succeed())
+  ```
+- Kubernetes apimachinery wait tooling.
+  Repeatedly executing the body function until the resulting assertion
+  succeeds, the timeout is reached or an error returned from the body.
+
+  Example:
+  ```
+  err := wait.Poll(time.Second*5, time.Minute*3, func() (bool, error) {
+      _, err := clusterApi.CoreV1().Secrets(namespace).Get(secret.Name, metav1.GetOptions{})
+      if err != nil {
+          if errors.IsNotFound(err) {
+              return true, nil
+          }
+          return false, nil
+      }
+      return false, fmt.Errorf("secret %s already exists", secret.Name)
+  })
+  ```
+- Kubevirt e2e wrapper for VMI state.
+  These are repeatedly reading the VMI state until the required
+  state/phase is reached. It does so using Ginkgo `Eventually`.
+  The required state condition is abstracted away.
+
+  Example:
+  ```
+  vmi = tests.CreateVmiOnNode(vmi, nodeName)
+  vmi = tests.WaitUntilVMIReady(vmi, console.LoginToFedora)
+  ```
+
+## Event driven waiting
+
+### Summary
+
+Kubernetes API is providing a "Watch" operation on all available
+resources. Watchers are a powerful toolset that enable [efficient
+detection of changes](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes)
+with assurance that no event is missed.
+
+Fortunately, the Kubernetes community has curated
+[tooling](https://pkg.go.dev/k8s.io/client-go/tools/watch) to take
+advantage of watchers by providing most of the mechanics required
+to setup and manage them.
+Its users are left with just adding basic logic regarding the
+conditions and possibly filtering (e.g. label selectors).
+
+The main function in use is `Until`, supporting a timeout, aborting
+and conditions.
+
+Example (waiting for namespace deletion):
+```
+_, err := k8swatchtools.Until(ctx, namespace.ResourceVersion, w, func(event k8swatch.Event) (bool, error) {
```

Comment:
this signature is wrong isn't it ?

Reply:
Can you help me detect the problem?
I took it from here: https://github.com/kiagnose/kiagnose/blob/3bb666abd01ddaf40941e7bd9f06443ef945e60f/kiagnose/internal/checkup/namespace/namespace.go#L68

Reply:
well, that's a call. I was expecting the signature (probably my bad, this is an example), which conveys more info. I have no idea what that `w` in there is - if I had the signature, I would know.

for convenience: https://github.com/kubernetes/client-go/blob/53da6da2bcb15dc987fa5b19eb33c4793e574c99/tools/watch/until.go#L106